417 add ConvertForMetrics post transform

Nic-Ma opened the pull request:
Fixes #417 .

### Description
This PR implemented ConvertForMetrics transform.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or new feature that would cause existing functionality to change)
- [x] New tests added to cover the changes
- [x] Docstrings/Documentation updated

Nic-Ma commented:
/black

wyli commented on monai/transforms/post/array.py:
```diff
@@ -46,3 +47,68 @@ def __call__(self, img, to_onehot=None, num_classes=None):
+        self.round_values = round_values
+        self.logit_thresh = logit_thresh
+
+    def __call__(self, y_pred, y=None):
```
I think this module does too many things, could decompose it into:
- `AsDiscrete`, for `argmax`, `to_one_hot`, `thresholding`, (round_values is not very accurate here)
- `NonLinear`, for `sigmoid`,`softmax`, other layers from `torch.nn`'s non linear functions collection https://pytorch.org/docs/stable/nn.html#non-linear-activations-other

also optionally we could have some basic input validation, e.g. sigmoid=True and softmax=True are not compatible options

Nic-Ma replied:
updated in the latest patch.
Thanks.

wyli commented on monai/transforms/post/array.py:
```diff
@@ -13,6 +13,7 @@
 https://github.com/Project-MONAI/MONAI/wiki/MONAI_Design
```
maybe add a comment here that this postprocessing module assumes shape `[batch, channels, (optional spatial dimensions)]`

also, [`write_nifti`](https://github.com/Project-MONAI/MONAI/blob/master/monai/data/nifti_writer.py#L19) `write_png` are also postprocessings, shall we move them here? (perhaps need to think about this, not a high priority task for now)

Nic-Ma replied:
updated in the latest patch.
`write_nifiti` can be also used by pre-processing transform, to save some transform result.
`save_nifiti` can be really post_transforms I feel.
You can submit an task issue for that I think.
Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @wyli ,

Thanks very much for your review.
I updated the code according to your comments.
Could you please help review it again?
Thanks.

Nic-Ma commented:
/black

wyli commented on monai/transforms/post/array.py:
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
+    def __call__(self, img, add_sigmoid=None, add_softmax=None, other=None):
+        if add_sigmoid is True and add_softmax is True:
+            raise ValueError("add_sigmoid=True and add_softmax=True are not compatible.")
+        if self.add_sigmoid if add_sigmoid is None else add_sigmoid:
```
`self.add_sigmoid if add_sigmoid is None else add_sigmoid` could be simplified as
`add_sigmoid or self.add_sigmoid`

Nic-Ma replied:
updated in the latest patch.
Thanks.

wyli commented on monai/transforms/post/array.py:
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
+
+    """
+
+    def __init__(self, add_sigmoid=False, add_softmax=False, other=None):
```
could replace `add_sigmoid` with `sigmoid`, `add_softmax` with `softmax`, and `AddActivations` with `Activations`, because I feel it's a bit confusing given we have `AddChannel`

Nic-Ma replied:
updated in the latest patch.
Thanks.

Nic-Ma commented on monai/transforms/post/array.py:
```diff
@@ -46,3 +48,72 @@ def __call__(self, img, to_onehot=None, num_classes=None):
+            img = torch.softmax(img, dim=1)
+        act_func = self.other if other is None else other
+        if act_func is not None:
+            if not isinstance(act_func, Callable):
```
updated in the latest patch.
Thanks.

Nic-Ma commented:
/black

Nic-Ma commented:
Hi @wyli ,

Thanks for your review.
I updated the code according to your comments.
Could you please help review it again?
I will also unify the `do_sigmoid`, `add_sigmoid`, `do_softmax`, etc. in DiceLoss, MeanDice and AUC metrics to `sigmoid`, `softmax` in another PR later.
Thanks.

wyli commented on tests/test_as_discreted.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class TestAsDiscreted(unittest.TestCase):
+    @parameterized.expand([TEST_CASE_2])
```
test cases not in use?

Nic-Ma replied:
updated in the latest patch.
Thanks.

wyli commented on docs/source/transforms.rst:
```diff
@@ -584,12 +597,25 @@ Dictionary-based Transforms
 
+`AddActivationsd`
+~~~~~~~~~~~~~~~~~
+.. autoclass:: AddActivationsd
```
now this needs to be changed as well

Nic-Ma replied:
updated in the latest patch.
Thanks.

Nic-Ma commented:
Hi @wyli ,

Thanks for your review, I updated the code.
Could you please help review it again?

wyli commented:
> Hi @wyli ,
> 
> Thanks for your review, I updated the code.
> Could you please help review it again?

thanks, this pr introduces a new warning for documentation building
```
MONAI/monai/transforms/post/array.py:docstring of monai.transforms.AsDiscrete:3: WARNING: Unexpected indentation.
```

Nic-Ma commented:
> > Hi @wyli ,
> > Thanks for your review, I updated the code.
> > Could you please help review it again?
> 
> thanks, this pr introduces a new warning for documentation building
> 
> ```
> MONAI/monai/transforms/post/array.py:docstring of monai.transforms.AsDiscrete:3: WARNING: Unexpected indentation.
> ```

Thanks for your reminder. Fixed in the latest patch.
